mne_sandbox/preprocessing/dss.py

```diff
@@ -0,0 +1,140 @@
+        non-zero values; to keep all values, specify ``thresh=None``.
+    bias_max_components : int | None
+        Maximum number of components to keep during PCA decomposition of the
+        bias function. ``None`` (the default) keeps all suprathreshold
+        components.
+    bias_thresh : float | None
+        Threshold (relative to the largest component) below which components
+        will be discarded during decomposition of the bias function. ``None``
+        (the default) keeps all non-zero values; to keep all values, pass
+        ``thresh=None`` and ``max_components=None``.
+    return_data : bool
+        Whether to return the denoised data along with the denoising matrix.
+
+    Returns
+    -------
+    dss_mat : array of shape (n_channels, n_dss_channels)
+        The denoising matrix.
+    dss_data : array of shape (n_trials, n_dss_channels, n_times)
+        The denoised data. Note that n_dss_channels are orthogonal virtual
+        channels and may be fewer in number than the number of channels in the
+        input Epochs object. Returned only if ``return_data`` is ``True``.
+
+    References
+    ----------
+    .. [1] Särelä, Jaakko, and Valpola, Harri (2005). Denoising source
+    separation. Journal of Machine Learning Research 6: 233–72.
+
+    .. [2] de Cheveigné, Alain, and Simon, Jonathan Z. (2008). Denoising based
+    on spatial filtering. Journal of Neuroscience Methods, 171(2): 331-339.
+    """
+    if isinstance(data, Epochs):
+        data_cov = compute_covariance(data).data
+        bias_cov = np.cov(data.average().data)
+        if return_data:
+            data = data.get_data()
+    elif isinstance(data, np.ndarray):
+        data_cov = np.sum([np.dot(trial, trial.T) for trial in data], axis=0)
+        bias_cov = np.cov(data.mean(axis=0))
+    else:
+        raise TypeError('Data to denoise must be an instance of mne.Epochs or '
+                        'a numpy array.')
+    dss_mat = _dss(data_cov, bias_cov, data_max_components, data_thresh,
+                   bias_max_components, bias_thresh)
+    if return_data:
+        n_trials, n_chans, n_times = data.shape
+        dss_data = np.einsum('ij,hik->hjk', dss_mat, data)
+        return dss_mat, dss_data
+    else:
+        return dss_mat
+
+
+def _dss(data_cov, bias_cov, data_max_components=None, data_thresh=None,
+         bias_max_components=None, bias_thresh=None):
+    """Process physiological data with denoising source separation (DSS)
+
+    Acts on covariance matrices; allows specification of arbitrary bias
+    functions (as compared to the public ``dss`` function, which forces the
+    bias function to be the evoked response).
+    """
+    data_eigval, data_eigvec = _pca(data_cov, data_max_components, data_thresh)
+    W = np.sqrt(1 / data_eigval)  # diagonal of whitening matrix
+    # bias covariance projected into whitened PCA space of data channels
+    bias_cov_white = (W * data_eigvec).T.dot(bias_cov).dot(data_eigvec) * W
```

Comment:
Did you do an e.g. `np.testing.assert_allclose` of this new version compared to your old one? I'm a little bit surprised that no `W[:, np.newaxis]` appears here but maybe that's because it implicily lives in the starting `(W * dat_eigvec).T`